kernel/src/engine/default/file_stream.rs

```diff
@@ -86,24 +90,77 @@ pub struct FileStream<F: FileOpener> {
     /// is not capable of limiting the number of records in the last batch, the file
     /// stream will take care of truncating it.
-    file_reader: F,
+    file_opener: Box<dyn FileOpener>,
     /// The stream state
     state: FileStreamState,
     /// Describes the behavior of the `FileStream` if file opening or scanning fails
     on_error: OnError,
 }
 
-impl<F: FileOpener> FileStream<F> {
-    /// Create a new `FileStream` using the give `FileOpener` to scan underlying files
+impl FileStream {
+    /// Creates a new `FileStream` from a given schema, `FileOpener`, and files list; the files are
+    /// processed asynchronously by the provided `TaskExecutor`. Returns an `Iterator` that consumes
+    /// the results.
+    pub fn new_async_read_iterator<E: TaskExecutor>(
+        task_executor: Arc<E>,
+        schema: ArrowSchemaRef,
+        file_opener: Box<dyn FileOpener>,
+        files: &[FileMeta],
+        readahead: usize,
+    ) -> DeltaResult<FileDataReadResultIterator> {
+        let mut stream = FileStream::new(files.to_vec(), schema, file_opener)?;
+
+        // This channel will become the output iterator
+        // The stream will execute in the background, and we allow up to `readahead`
+        // batches to be buffered in the channel.
+        let (sender, receiver) = std::sync::mpsc::sync_channel(readahead);
+
+        let executor_for_block = task_executor.clone();
+        task_executor.spawn(async move {
+            while let Some(res) = stream.next().await {
+                let sender = sender.clone();
+                let join_res = executor_for_block
+                    .spawn_blocking(move || sender.send(res))
+                    .await;
+                match join_res {
+                    Ok(send_res) => match send_res {
+                        Ok(()) => continue,
+                        Err(_) => break,
+                    },
+                    Err(je) => {
+                        panic!("Couldn't join spawned task, runtime is likely in bad state: {je}")
+                    }
+                }
+            }
+        });
+
+        // Create a thread-safe iterator, because engine may consume it from multiple threads.
+        //
+        // NOTE: Every call to the iterator's `next` method will lock the mutex before accessing the
+        // underlying stream. This should not be a bottleneck because the stream will immediately
+        // return an item (if ready) and would anyway block if not ready. Additionally, each
+        // iterator element corresponds to a file access whose cost dwarfs any mutex overhead.
+        let it = receiver
+            .into_iter()
+            .map(|rbr| rbr.map(|rb| Box::new(ArrowEngineData::new(rb)) as _));
+        let it = std::sync::Mutex::new(it);
+        let it = std::iter::from_fn(move || match it.lock() {
+            Ok(mut i) => i.next(),
+            Err(_) => Some(Err(Error::generic("Poisoned mutex"))),
+        });
```

Comment:
This is the bug fix -- previously we just returned the receiver iterator directly.